linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1150,6 +1186,119 @@ public void close() {
     };
   }
 
+  /** Implementation of join that builds the right input and probes with the
+   * left */
+  private static <TSource, TInner, TKey, TResult> Enumerable<TResult> hashJoinWithPredicate_(
+      final Enumerable<TSource> outer, final Enumerable<TInner> inner,
+      final Function1<TSource, TKey> outerKeySelector,
+      final Function1<TInner, TKey> innerKeySelector,
+      final Function2<TSource, TInner, TResult> resultSelector,
+      final EqualityComparer<TKey> comparer, final boolean generateNullsOnLeft,
+      final boolean generateNullsOnRight, final Predicate2<TSource, TInner> predicate) {
+
```

Comment:
I think the first line of this method could be:
`if (predicate == null) return hashEquiJoin_(...);`
That should help performance in that case, and would also help simplify the code of the `hashJoinWithPredicate_` method (no need to check `predicate == null` / `predicate != null` )

Reply:
done